pkg/image/controller/controller.go

```diff
@@ -29,6 +123,73 @@ func needsImport(stream *api.ImageStream) bool {
 	return stream.Annotations == nil || len(stream.Annotations[api.DockerImageRepositoryCheckAnnotation]) == 0
 }
 
+func (c *ImportController) handleImport() {
+	for {
+		select {
+		case <-c.stopChan:
+			return
+
+		case staleImageStream := <-c.work:
+			glog.V(1).Infof("popped %s from the worklist", workingSetKey(staleImageStream))
+
+			// if we're already in the workingset, that means that some thread is already trying to do an import for this.
+			// This does NOT mean that we shouldn't attempt to do this work, only that we shouldn't attempt to do it now.
+			if c.isInWorkingSet(staleImageStream) {
+				// If there isn't any other work in the queue, wait for a while so that we don't hot loop.
+				// Then requeue to the end of the channel.  That allows other work to continue without delay
+				if len(c.work) == 0 {
+					time.Sleep(100 * time.Millisecond)
+				}
+				glog.V(5).Infof("requeuing %s to the worklist", workingSetKey(staleImageStream))
+				c.work <- staleImageStream
+			}
+
+			c.addToWorkingSet(staleImageStream)
+
+			err := kclient.RetryOnConflict(kclient.DefaultBackoff, func() error {
+				liveImageStream, err := c.streams.ImageStreams(staleImageStream.Namespace).Get(staleImageStream.Name)
+				if err != nil {
+					return err
+				}
+				if !needsImport(liveImageStream) {
+					return nil
+				}
+
+				// if we're notified, do work and then start waiting again.
+				return c.Next(liveImageStream)
+			})
+
+			c.removeFromWorkingSet(staleImageStream)
+			if err != nil {
+				util.HandleError(err)
+			}
+
+		}
+	}
+}
+
+func workingSetKey(imageStream *api.ImageStream) string {
+	return imageStream.Namespace + "/" + imageStream.Name
```

Comment:
> Instead of creating one, why not using controller.KeyFunc from pkg/controller?

This is non-reflective and type-safe.  The keyfuncs for a store aren't.